arch/rv32i/src/pmp.rs

```diff
@@ -490,210 +683,472 @@ impl<const MAX_AVAILABLE_REGIONS_OVER_TWO: usize> kernel::platform::mpu::MPU
+                        csr::CSR.pmpaddr_set(i * 2 + 3, odd_region.2 as usize);
+                    }
 
-    fn enable_kernel_mpu(&self, config: &mut Self::KernelMpuConfig) {
-        for (i, region) in config.regions.iter().rev().enumerate() {
-            let x = MAX_AVAILABLE_REGIONS_OVER_TWO - i - 1;
-            match region {
-                Some(r) => {
-                    let cfg_val = r.cfg.value as usize;
-                    let start = r.location.0 as usize;
-                    let size = r.location.1;
-
-                    match x % 2 {
-                        0 => {
-                            csr::CSR.pmpaddr_set((x * 2) + 1, (start + size) >> 2);
-                            // Disable access up to the start address
-                            csr::CSR.pmpconfig_modify(
-                                x / 2,
-                                csr::pmpconfig::pmpcfg::r0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::w0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::x0::CLEAR
-                                    + csr::pmpconfig::pmpcfg::a0::CLEAR,
-                            );
-                            csr::CSR.pmpaddr_set(x * 2, start >> 2);
-
-                            // Set access to end address
-                            let new_cfg =
-                                cfg_val << 8 | (csr::CSR.pmpconfig_get(x / 2) & 0xFFFF_00FF);
-                            csr::CSR.pmpconfig_set(x / 2, new_cfg);
-                            // Lock the CSR
-                            csr::CSR.pmpconfig_modify(x / 2, csr::pmpconfig::pmpcfg::l1::SET);
-                        }
-                        1 => {
-                            csr::CSR.pmpaddr_set((x * 2) + 1, (start + size) >> 2);
-                            // Disable access up to the start address
-                            csr::CSR.pmpconfig_modify(
-                                x / 2,
-                                csr::pmpconfig::pmpcfg::r2::CLEAR
-                                    + csr::pmpconfig::pmpcfg::w2::CLEAR
-                                    + csr::pmpconfig::pmpcfg::x2::CLEAR
-                                    + csr::pmpconfig::pmpcfg::a2::CLEAR,
-                            );
-                            csr::CSR.pmpaddr_set(x * 2, start >> 2);
-
-                            // Set access to end address
-                            let new_cfg =
-                                cfg_val << 24 | (csr::CSR.pmpconfig_get(x / 2) & 0x00FF_FFFF);
-                            csr::CSR.pmpconfig_set(x / 2, new_cfg);
-                            // Lock the CSR
-                            csr::CSR.pmpconfig_modify(x / 2, csr::pmpconfig::pmpcfg::l3::SET);
-                        }
-                        // unreachable, but don't insert a panic
-                        _ => (),
+                    i += 2;
+                } else {
+                    // TODO: check overhead of code
```

Comment:
TODO: Analyze overhead of the implemented PMP configuration routine. Initial results seem to suggest that this is much more efficient than before, but calling out to the `pmpconfig_*` methods is still very costly. We might be able to cut down on runtime cost by using a hand-written ASM jump table for these functions.